core/src/com/unciv/ui/worldscreen/unit/UnitActions.kt

```diff
@@ -12,381 +12,397 @@ import com.unciv.models.UncivSound
+    private suspend fun SequenceScope<UnitAction>.yieldDisbandAction(unit: MapUnit, worldScreen: WorldScreen) {
+        yield(UnitAction(
                 type = UnitActionType.DisbandUnit,
-                canAct = unit.currentMovement > 0,
-                action = {
+                action = getLambdaOrNull(unit.currentMovement > 0) {
                     val disbandText = if (unit.currentTile.getOwner() == unit.civInfo)
                         "Disband this unit for [${unit.baseUnit.getDisbandGold()}] gold?".tr()
                     else "Do you really want to disband this unit?".tr()
                     YesNoPopup(disbandText, { unit.disband(); worldScreen.shouldUpdate = true }).open()
-                })
+                }))
     }
 
-    private fun addCreateWaterImprovements(unit: MapUnit, tile: TileInfo, actionList: ArrayList<UnitAction>) {
-        for (improvement in listOf("Fishing Boats", "Oil well")) {
-            if (unit.hasUnique("May create improvements on water resources") && tile.resource != null
-                    && tile.isWater // because fishing boats can enter cities, and if there's oil in the city... ;)
-                    && tile.improvement == null
-                    && tile.getTileResource().improvement == improvement
-                    && unit.civInfo.tech.isResearched(unit.civInfo.gameInfo.ruleSet.tileImprovements[improvement]!!.techRequired!!)
-            )
-                actionList += UnitAction(
-                        type = UnitActionType.Create,
-                        title = "Create [$improvement]",
-                        canAct = unit.currentMovement > 0,
-                        action = {
-                            tile.improvement = improvement
-                            unit.destroy()
-                        })
-        }
+    private suspend fun SequenceScope<UnitAction>.yieldCreateWaterImprovements(tile: TileInfo,
+                                                                               unit: MapUnit) {
+        val improvement = tile.getTileResourceOrNull()?.improvement
+        if (improvement != null && tile.isWater && tile.improvement == null
+                && unit.hasUnique("May create improvements on water resources")
+                && unit.civInfo.tech.isResearched(
+                        unit.civInfo.gameInfo.ruleSet.tileImprovements[improvement]!!
+                                .techRequired!!))
+            yield(UnitAction(
+                    type = UnitActionType.Create,
+                    title = "Create [$improvement]",
+                    action = getWaterImprovementAction(unit, tile, improvement)))
     }
 
-    private fun addConstructRoadsAction(unit: MapUnit, tile: TileInfo, actionList: ArrayList<UnitAction>) {
-        if (unit.hasUnique("Can construct roads")
-                && tile.roadStatus == RoadStatus.None
+    fun getWaterImprovementAction(unit: MapUnit, tile: TileInfo, improvement: String): (() -> Unit)? =
```

Comment:
this is called elsewhere to directly get the water improvement action rather than doing the search of list of all possible actions